writer.go

```diff
@@ -0,0 +1,145 @@
+package pinecone
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/pinecone-io/go-pinecone/pinecone"
+	"github.com/pkg/errors"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+// Writer implements a writer logic for Sap hana destination.
+type Writer struct {
+	client *pinecone.Client
+	index  *pinecone.IndexConnection
+}
+
+// NewWriter New creates new instance of the Writer.
+func NewWriter(ctx context.Context, config DestinationConfig) (*Writer, error) {
+	sdk.Logger(ctx).Trace().Msg("Creating new writer.")
+
+	sdk.Logger(ctx).Error().Msgf("API: %v, INDEX:%v", config.PineconeAPIKey, config.PineconeHostURL)
+
+	pineconeClient, indexConnection, err := NewPineconeClient(ctx, config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create pinecone pineconeClient: %v", err)
+	}
+
+	writer := &Writer{
+		client: pineconeClient,
+		index:  indexConnection,
+	}
+	return writer, nil
+}
+
+func (w *Writer) Upsert(ctx context.Context, record sdk.Record) error {
+	ID := recordID(record.Key)
+
+	payload, err := recordPayload(record.Payload)
+	if err != nil {
+		return fmt.Errorf("error getting payload: %v", err)
+	}
+
+	metadata, err := recordMetadata(record.Metadata)
+	if err != nil {
+		return fmt.Errorf("error getting metadata: %v", err)
+	}
+
+	sdk.Logger(ctx).Error().Msgf("metadata: %v", metadata)
+
+	vector := []*pinecone.Vector{{
+		Id:     ID,
+		Values: payload,
+		SparseValues: nil,
+		Metadata:     metadata,
+	}}
+
+	_, err = w.index.UpsertVectors(&ctx, vector)
+	if err != nil {
+		return fmt.Errorf("error upserting record: %v ", err)
+	}
+
+	sdk.Logger(ctx).Trace().Msg("upserted record")
+	return nil
+}
+
+// Delete deletes records by a key.
+func (w *Writer) Delete(ctx context.Context, record sdk.Record) error {
+	ID := []string{recordID(record.Key)}
+
+	err := w.index.DeleteVectorsById(&ctx, ID)
+	if err != nil {
+		return fmt.Errorf("\n error deleting record: %v ", err)
```

Comment:
Unless there's a special reason for the newline at the beginning of the error message, we can remove it.

Reply:
Also refactored that in another pr.